Provide Distributed Switch Architecture [DSA] support (based on ip_k66f board)

lmajewski opened the pull request:
**Description:**
In this PR the support for DSA [Distributed Switch Architecture] on ip_k66f
board (which uses NXP's K66F MCU and KSZ8794 Switch IC) is provided.

To be more specific - the LLDP protocol handling is used to show lan1, lan2 and
lan3 ports separation.

When ip_k66f receives LLDP packet with destination MAC 01:80:c2:00:00:03, it is
parsed and displayed on the console.
Afterwards, the board replies with its own LLDP information via the LAN port
on which it has received this packet.

The port separation is only configured for 01:80:c2:00:00:03 MAC address - packets
with other addresses works correctly - i.e. packets are passed between lan{123}
ports and MCU. This means that other protocols like ICMP, TCP/IP works on any
lan{123} ports.

**Note:**
It was necessary to modify in some points the generic network stack code.

The previous RFC for DSA handling (with some previous discussion)
can be found here:
https://github.com/zephyrproject-rtos/zephyr/pull/25008

**Build procedure:**
`west build --pristine -b ip_k66f ./zephyr/samples/net/dsa/ && west debug`

**The test procedure:**

On HOST - configure/start the lldpd daemon on host to use MAC 01:80:c2:00:00:03:
`(lldpd -ddd & ) && sleep 2 && lldpcli configure lldp agent-type nearest-non-tpmr-bridge`

The above line shall be executed on at least two PCs connected to lan{123} of
the ip_k66f board.

Commit:
drivers: ethernet: Add support for KSZ8794 DSA device

This patch adds support for Microchip's KSZ8794 DSA device, which for
switch and PHY control uses SPI communication.

This driver also provides support for handling tail tagging added and
decoded in the KSZ8794 device as well as modifying entries in the
static MAC address table.

It is also possible to use GPIO pin to reset KSZ8794 switch.
When the "reset-gpios" property is not defined, software based reset
is performed instead.

The KSZ8794 can now be used on boards which have SPI CS only available
as GPIO pin.

Signed-off-by: Stefan Bigler <stefan.bigler@securiton.ch>
Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
eth: Add support for Distributed Switch Architecture [DSA] switches

This patch add support for DSA switches to Zephyr.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
eth: mcux: Add support for Distributed Switch Architecture devices

This patch adds support for DSA devices - like ksz8794 switch to the
mcux Kinetics driver.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
samples: net: Provide DSA example to handle LLDP MGNT packets on ip_k66f

This patch brings simple example on per DSA port LLDP filtering based on
MAC address of received packet.

Each lan1, lan2 and lan3 ports handle separately LLDP packets going into
them.

Nonetheless, the K66F ENET handles correctly all other packets types -
to be more specific - ICMP, TCP/IP.

Setup -> ON HOST:

Configure/start the lldpd daemon (debian 10) on host to use
MAC 01:80:c2:00:00:03:
(lldpd -ddd & ) && sleep 2 && \
lldpcli configure lldp agent-type nearest-non-tpmr-bridge

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dsa: dts: Add microchip,ksz8794.yaml file describing DSA switch properties

The ksz8794 yaml file provides information regarding device tree's
properties for this particular DSA switch IC.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dsa: config: sample: Add SHELL support via RTT to DSA sample's prj.conf

Due to the routing, the ip_k66f board can only use RTT to export
console. With this change the SHELL is enabled to get access
to network commands (like ping).

To use it with RTT from Segger:
-------------------------------

On HOST (terminal 1):
./JLink_V664/JLinkRTTLogger -Device MK66FN2M0XXX18 -RTTChannel 1 \
-if SWD -Speed 4000 ~/rtt.log

On HOST (terminal 2):
nc localhost 19021

(19021 is the port number for the Segger RTT server)

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dsa: ip_k66f: Enable support for KSZ8794 DSA device on ip_k66f board

This change enables support for KSZ8794 DSA device on the ip_k66f
board. Each LAN port is defined as a DTS subnode.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

tbursztyka commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,579 @@
+		dsa_ksz8794_write_reg(sdev, KSZ8794_CTRL2_PORTn(i), tmp);
+	}
+
+#if IS_ENABLED(CONFIG_DSA_KSZ8794_TAIL_TAGGING)
```
We use IS_ENABLED() for C if, not pre-processor #if. 
(unless it's accepted to use IS_ENABLED, but the purpose of that macros was to avoid pre-processor #if )

Used defined() here. Or switch to an actual C if/else with IS_ENABLED

lmajewski replied:
Ok. Thanks for pointing this out.

tbursztyka commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,579 @@
+
+	buf = net_buf_alloc_len(net_buf_pool_get(pkt->buffer->pool_id),
+				DSA_KSZ8794_INGRESS_TAG_LEN, K_NO_WAIT);
+	buf = net_buf_frag_last(net_buf_frag_add(pkt->buffer, buf));
```
you could avoid allocating 2 net_buf here, just allocate one at the beginning of the function (ask for the max length you need, it's anyway going to fit in actual fixed length of net_buf). Pre-fill it with padding if necessary, then with port_idx.

lmajewski replied:
The code as is looks more logical to me as:

1. I first check if the packet is smaller than 64B (if yes - I allocate padding and append it)
2. Then I do allocate always (no matter of the packet size) a single byte for the tag, which logically is not related to preparing buffer for packet, as it is removed in the KSZ8794 switch.
3. The last step is to setup the tag byte to the buffer allocated earlier.

The code as it is works seamlessly when the packet is larger than 64B (minimal packet size), as we would need the extra allocation for tag always. For packet smaller than 64B we would have two buffers allocated and appended (pad and tag) to the original packet.

tbursztyka replied:
It can be kept logical _and_ memory efficient. Here you'll eat up 256 bytes for at most 65bytes (if net_buf buffer size is 128bytes, it gets worse if that value is higher: 512 is 256 etc etc...).

- calculate the padding
- allocate for the padding and the tag
- if (pad) ... add the padding
- add the tag.

lmajewski replied:
Ok, I've rewritten it to use only single `net_buf_alloc_len()` allocation. Thanks for sharing the idea.

jukkar commented on include/net/dsa.h:
```diff
@@ -0,0 +1,283 @@
+#define DSA_PORT3_MAC DT_PROP(DSA_DT_NODE, local_mac_address_port3)
+#endif
+
+#define DSA_NET_SLAVE_PORTS_INIT()	  \
```
This is now mcux specific, should we place this to mcux driver header file in `drivers/ethernet/` directory instead?

lmajewski replied:
This `#define DSA_NET_SLAVE_PORTS_INIT()` uses generic `NET_DEVICE_INIT_INSTANCE()` to instantiate network interfaces. Other functions - like reading byte from DSA IC looks to be generic. Hence the idea to store it in dsa.h.

jukkar replied:
Sure, but the macro uses MCUX specific settings. One option is to make them as parameters to this macro.
For example, the instance of the context variable could be a parameter to this macro, same for ETH_MCUX_PM_FUNC .

lmajewski replied:
Or I can remove them and just use the MAC address assigned for the eth0 (master) interface.

lmajewski replied:
Please find updated PR code. It now is more clear and has been moved to dsa_ksz8794.c driver file.

tbursztyka commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,579 @@
+	net_pkt_read_u8(pkt, &pnum);
+
+	net_pkt_update_length(pkt, plen - DSA_KSZ8794_EGRESS_TAG_LEN);
+	plen = net_pkt_get_len(pkt);
```
avoid calling net_pkt_get_len again here, just update plen accordingly. net_pkt_get_len is always going through the net_buf linked list and is thus not optimized

lmajewski replied:
I could remote this call from the line 517, as it is used only in the NET_DBG() macro for debugging.
It can be replaced wtih plen - DSA_KSZ8794_EGRESS_TAG_LEN in the NET_DBG() call.

jukkar commented on include/net/ethernet.h:
```diff
@@ -126,6 +139,10 @@ enum ethernet_hw_caps {
 	ETHERNET_HW_VLAN_TAG_STRIP	= BIT(14),
+
+	/** DSA switch */
+	ETHERNET_DSA_SLAVE_PORT	= BIT(15),
```
Indentation seems to be off here

lmajewski replied:
After applying the patch it looks to be consistent with the rest of this file (by using TABs).

tbursztyka commented on include/net/ethernet.h:
```diff
@@ -74,7 +78,16 @@ struct net_eth_addr {
+ * Extend the max frame size for DSA (KSZ8794) by one byte (to 1519) to
+ * store tail tag.
+ */
+#if IS_ENABLED(CONFIG_NET_DSA) && IS_ENABLED(CONFIG_DSA_KSZ8794)
```
ethernet.h should stay as generic as possible, remove the second IS_ENABLED()

lmajewski replied:
Ok.

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -974,6 +974,13 @@ static size_t pkt_buffer_length(struct net_pkt *pkt,
 		if (net_if_l2(net_pkt_iface(pkt)) ==
 		    &NET_L2_GET_NAME(ETHERNET)) {
 			max_len += sizeof(struct net_eth_hdr);
+#if IS_ENABLED(CONFIG_NET_DSA) && IS_ENABLED(CONFIG_DSA_KSZ8794)
```
same here

lmajewski replied:
Ok.

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -974,6 +974,13 @@ static size_t pkt_buffer_length(struct net_pkt *pkt,
 			max_len += sizeof(struct net_eth_hdr);
+#if IS_ENABLED(CONFIG_NET_DSA) && IS_ENABLED(CONFIG_DSA_KSZ8794)
+			/*
+			 * Extend the max pkt length for DSA (KSZ8794) by one
```
if such tag is found in all DSA, no need to tell about a specific one.

If not, then that is not a proper solution (the tag would need no stay always managed and hidden in drivers/ethernet/dsa)

lmajewski replied:
I've added following solution to the `dsa.h` file.

`#if defined(CONFIG_DSA_KSZ8794) && defined(CONFIG_DSA_KSZ8794_TAIL_TAGGING)`
`#define _DSA_TAG_SIZE 1`
`#else`
`#define _DSA_TAG_SIZE 0`
`#endif`

And then only use `_DSA_TAG_SIZE` in generic code.

lmajewski replied:
The above change has been applied to updated PR.

jukkar commented on samples/net/dsa/README.rst:
```diff
@@ -0,0 +1,31 @@
+.. _dsa-sample:
+
+DSA Sample Application
+#######################
```
There seems to be one extra `#` here

lmajewski replied:
Fixed.

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -393,6 +396,9 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	net_pkt_cursor_init(pkt);
 
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
```
not sure it's the right place to do this. dsa is ethernet specific, and as such maybe ethernet L2 should be doing this

lmajewski replied:
The idea here is to read LAN port number (lan1,2,3) as soon as possible - i.e. from the raw buffer before any high level network code is executed.
For example here we read the LAN interface number (from appended tag by KSZ8794) before the struct iface is set for the packet. IMHO this is the best conceptual place to avoid issues latter.

tbursztyka replied:
no architecturally this is by far the worst place: net_recv_data is a very generic function and you should not add L2 specific use cases there.

[edit] all DSA related stuff, besides the drivers,  must land in subsys/net/l2/ethernet

lmajewski replied:
The rationale to place `dsa_recv_set_iface()` function in net_core.c's `net_recv_data()` function was to set proper `struct iface` just before it is used in `net_pkt_set_iface()`. Afterwards the packet is passed to `net_queue_rx()`.

When moving the `dsa_recv_set_iface()` to `ethernet_recv()`, in the DSA case, we would have the iface set to master
port (which is NOT correct) when the packet enters `ethernet_recv()` (in net/l2/ethernet/ethernet.c). Would it be safe to have such mismatch?

jukkar replied:
If you set the interface properly already in the driver before calling `net_recv_data()`, there is no need to have any changes in net_core.c

lmajewski replied:
I agree. I've moved this code to `eth_rx()` callback in eth_mcux.c

lmajewski replied:
After some excessive testing it turned out that the dsa_recv_set_iface() shall
be called in net_recv_data() (subsys/net/ip/net_core.c) as after some basic
checks (if e.g. iface != NULL) we are sure that the iface is ready to handle
the packet.

Moreover, in the same place the code to assign original iface (when
CONFIG_NET_ROUTING is enabled) is executed and hence the deduction that
DSA shall follow the same pattern).

tbursztyka commented on drivers/ethernet/Kconfig:
```diff
@@ -55,5 +55,6 @@ source "drivers/ethernet/Kconfig.native_posix"
 source "drivers/ethernet/Kconfig.stellaris"
```
split this patch in parts: 
- one that add the dsa core part (include/net/ and drivers/ethernet/dsa)
- one for net core specifics

lmajewski replied:
Could you be more specific here?
I do guess that you refer to: `eth: Add support for Distributed Switch Architecture [DSA] switches` patch.
How shall it be split?

tbursztyka commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,579 @@
+
+	/* SPI config */
+	sdev->spi_cfg.operation =
+#if DSA_DT_SPI_CPOL
```
shouldn't you be able to get these from DTS?

lmajewski replied:
The DSA_DT_SPI_CPOL is already set according to DTS. Marcros for it can be found here:
`include/net/dsa.h`

lmajewski replied:
Please find updated version of this patch set - it now uses `DT_DRV_COMPAT`

jukkar commented on samples/net/dsa/src/main.h:
```diff
@@ -0,0 +1,117 @@
+#define DSA_PRIORITY      5
+#define DSA_THREAD_START_DELAY 4000
+
+#define DSA_THREAD(ID, FN_RECV, FN_SEND)                                       \
```
This certainly works but could be simplified by using `UTIL_LISTIFY()` macro, see for example how it is used in `drivers/console/uart_mux.c`. No need to change this now, but just an idea for future development.

lmajewski replied:
Thanks for the tip.

jukkar reviewed (commented):
Looks quite resonable, some minor comments.

tbursztyka commented on subsys/net/ip/connection.c:
```diff
@@ -578,8 +578,12 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
+		if ((IS_ENABLED(CONFIG_NET_UDP) ||
+		     IS_ENABLED(CONFIG_NET_TCP))
+#if IS_ENABLED(CONFIG_NET_DSA)
+		    && !net_dsa_recv(pkt_iface, pkt)
```
I don't think it should hit connection.c ever.
exactly as in your modification in net_core.c: this is not the right place.

When an ethernet packet is received it goes to net_core which in turn will push it to its related L2. So if something has to be done at low level with such packet, it's in ethernet L2.

lmajewski replied:
The rationale for this change:

The Zephyr eth stack is not (by default) prepared to work with RAW and TCP/UDP sockets enabled in the same time.
In this case the decision about forwarding is based on the destination MAC address. If we find MAC, which we are looking for (LLDP MAC: 01:80:c2:00:00:03), then the TCP/UDP stack is not used.

Without this patch it would NOT be possible to:
1. Forward all network traffic (ICMP, TCP/IP)  to ETH_MCUX master - (no matter to which port it was delivered)
2. Handle RAW packets - like LLDP only with lan1,2,3 interfaces.

jukkar replied:
AF_PACKET with SOCK_RAW has some limitations atm when working with TCP/UDP sockets. AF_PACKET with SOCK_DGRAM support was recently merged which allows Ethernet L2 to create Ethernet frame and user to provide rest of the data. These kind of AF_PACKET sockets will work much better with other TCP/UDP sockets. Dunno if that would help here.
Anyway, I agree with @tbursztyka here, we should think harder about how the data is moving around in the network stack. Now we are exposing and calling bearer/driver specific functions from generic code which does not look good.

lmajewski replied:
@jukkar The problem is that I would prefer to use `AF_PACKET` and `SOCK_RAW`. 
@jukkar @tbursztyka  As I've shown in this patch series - [some hacks are necessary](https://github.com/lmajewski/zephyr/commit/c0e85e4cc102a5e762a5b6094f5779b55d7a5583) to get `AF_PACKET` and `SOCK_RAW`

IMHO, as Zephyr is supposed to work with embedded/resources constrained systems, it would be welcome for users to allow them to craft frames in a way they prefer.

jukkar replied:
> IMHO, as Zephyr is supposed to work with embedded/resources constrained systems, it would be welcome for users to allow them to craft frames in a way they prefer.

Sure, for raw sockets that is expected that the application needs to create the packet itself anyway. I think no one is trying to prevent that.

lmajewski replied:
@tbursztyka @jukkar - To ask in another way - what would you recommend to have AF_PACKET and SOCK_RAW with DSA? Is it possible to accept mentioned in my last reply "hack" to Zephyr? If not - what shall be done to accept it?

lmajewski replied:
@tbursztyka @jukkar Is there any new approach for the AF_PACKET and SOCK_RAW support? Could you give me some advice on the way I shall proceed?

tbursztyka reviewed (changes requested):
code is very clean
just some comments, and one architectural issue I believe

lmajewski commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,590 @@
+{
+	uint8_t tmp;
+
+	if (port < KSZ8794_PORT1 || port > KSZ8794_PORT3)
```
Fixed.

lmajewski replied:
Hmm, strange - I was almost 100% sure that I've fixed it already. Thanks for double checking.

lmajewski commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,590 @@
+					    const uint8_t *mac, uint8_t fw_port,
+					    uint16_t entry_idx)
+{
+	uint8_t buf[8];
```
Fixed.

lmajewski replied:
Ok, I've sorted this out - I've rebased on branch, which did not have those fixes applied. I will fix it globally and re-send the PR.

lmajewski commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,590 @@
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Fixed.

lmajewski replied:
The same commit as above.

lmajewski commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+
+int net_dsa_recv(struct net_if *iface, struct net_pkt *pkt)
+{
+
```
Fixed.

tbursztyka commented on subsys/net/ip/net_pkt.c:
```diff
@@ -974,6 +974,10 @@ static size_t pkt_buffer_length(struct net_pkt *pkt,
 		if (net_if_l2(net_pkt_iface(pkt)) ==
 		    &NET_L2_GET_NAME(ETHERNET)) {
 			max_len += sizeof(struct net_eth_hdr);
+#if defined(CONFIG_NET_DSA)
```
I would avoid such ifdef here.
Instead in include/ethernet.h create a macro like NET_ETH_MAX_HDR_SIZE  which is by default sizeof(struct net_eth_hdr)
if DSA is enabled: add +1 (have a macro for the DSA tag size as well).

Finally here, replace sizeof(struct net_eth_hdr) by NET_ETH_MAX_HDR_SIZE

lmajewski replied:
Ok. Thanks for the idea. I've rewritten the code.

tbursztyka commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+	return iface;
+}
+
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt)
```
looking at this, I really don't see why it could not land in ethernet_recv()

lmajewski replied:
Please find explanation with the comment with `net_core.c` where this function is used.

tbursztyka commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+	 * not for the master interface, which shall support all other
+	 * protocols - i.e. UDP. ICMP, TCP.
+	 */
+	if (net_dsa_recv(iface_sw, *pkt))
```
{ } missing

lmajewski replied:
Ok.

lmajewski commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -710,7 +715,7 @@ static bool eth_get_ptp_data(struct net_if *iface, struct net_pkt *pkt,
 #endif /* CONFIG_PTP_CLOCK_MCUX */
 
-static int eth_tx(struct device *dev, struct net_pkt *pkt)
+int eth_tx(struct device *dev, struct net_pkt *pkt)
```
Yes, correct. This function is called to reuse code to transmit data with master interface (after doing some adjustments).
In fact, this is non-elegant reuse of `ethernet_api.send` method (eth_tx).

lmajewski replied:
Would it be possible to devise the roadmap for having AF_PACKET and SOCK_RAW supported in the network stack?
It is crucial for the use case which is utilizing the DSA framework.

lmajewski commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -0,0 +1,185 @@
+#include "main.h"
+
+/* Loglevel of dsa_lldp function */
+LOG_MODULE_REGISTER(dsa_lldp, LOG_LEVEL_DBG);
```
Ok.

jukkar replied:
Please fix

lmajewski commented on samples/net/dsa/src/main.c:
```diff
@@ -0,0 +1,129 @@
+	dsa_set_mac_table_entry(&eth_filter_l2_addr_base[0][0], BIT(4), 0, 0);
+	dsa_get_mac_table_entry(tbl_buf, 0);
+
+	LOG_INF("KSZ8794 static MAC address table entry [%d]:", 0);
```
Ok,

lmajewski commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -0,0 +1,185 @@
+		  struct eth_addr *origin_addr)
+{
+	int ret, len;
+	char buffer[DSA_BUF_SIZ], s[24];
```
The `s` is a shorthand for string table. It is used to provide LLDP information about "System Name".
I've adjusted it properly.

jukkar replied:
Please fix.

lmajewski commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_dsa);
```
Fixed. Thanks for the idea.

lmajewski commented:
I would like to have the "stale" label removed as I cannot do it by myself.

lmajewski commented:
Dear Zephyr community (@tbursztyka @jukkar @pfalcon ),

Please find updated version of DSA support (KSZ8794).
The most notable changes:
- Add workarounds for this Switch - handle the issue with short cables (the in-depth documentation is in the dsa_ksz8794.c)
- Add support for hard reset (pin  GPIO PTA24) - this is an optional feature
- Add software/gpio controlled CS for the switch - optional feature
- Adjust the code to the newest master - also providing some fixes (https://github.com/zephyrproject-rtos/zephyr/pull/30428)

**Open points to discuss:**
-  Reusing of `eth_tx` from `dsa_tx` in eth_mcux.c
https://github.com/zephyrproject-rtos/zephyr/pull/27620/commits/520eedc3243bd3a6f34ef91869a6ee07ca560edc#diff-7c35b98f45f82570005c7d3bc596b8a35e30577a1a7d151b1989370e3d6cd457R1134
I should wrap it to some generic interface - ideas more than welcome

- Add proper support for `SOCK_RAW` and `AF_PACKETS` - is there any plan for such support? (If not - I can work on it but, I would need some guidance on how to efficiently fit it into Zephyr's API)

jukkar commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+	return 0;
+}
+
+__weak int dsa_set_mac_table_entry(const uint8_t *mac, uint8_t fw_port,
```
You can extend ethernet_api and put these functions there. Then just provide a function like this, you just need to add iface parameter to this call so we can figure out the api to use, and then call the actual function provided by the driver.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,169 @@
+/*
+ * TRANSMISSION HANDLING CODE egress (DSA slave ports -> ETH)
+ */
+int eth_tx(const struct device *dev, struct net_pkt *pkt);
```
No, we do not do things like this,  calling an api from driver which then calls back the driver using a global name. Perhaps you could call `dsa_tx()` from driver and then give a function callback as a parameter here. Then the functions in the driver could be kept private.

jukkar commented on samples/net/dsa/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
```
In commit `dsa: example: ...`, please change the commit subject to `samples: net: dsa: ....`

lmajewski replied:
Ok.

jukkar commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -0,0 +1,185 @@
+	uint8_t *p, t1, t2;
+	char *t;
+
+	LOG_INF("LLDP pkt recv -> lan%d", lanid);
```
Use `LOG_DBG` as this is clearly a debug msg (and we have dbg level set anyway)
Same things to INF calls below.

lmajewski replied:
If you don't mind - I would opt for keeping this,as it is part of the example/test. 

With `LOG_INF` I can see the MAC address of received packet and print information from which KSZ8794 LAN port the LLDP packet has been read.
This allows me to check (with e.g. wireshark/tcpdump) if the "port separation" feature (as configured in this example) works as expected.
Moreover, with this information I can check if board did not hang.

jukkar replied:
> If you don't mind - I would opt for keeping this,as it is part of the example/test.

No worries, we can keep the INF level.

jukkar commented on samples/net/dsa/src/dsa_lldp.c:
```diff
@@ -0,0 +1,185 @@
+			LOG_INF("\tTTL:\t\t%ds", (uint16_t) p[0] << 8 | p[1]);
+			break;
+		case LLDP_TLV_SYSTEM_NAME:
+			t = malloc(length+1);
```
Looks like an overkill to use malloc() here. You could use a buffer with some sane length instead.

lmajewski replied:
Here I would also prefer to keep `malloc`. I do know that it is the overkill at least in this example, but any "real" use case  (and hence code) would use `malloc`.

pfalcon replied:
> but any "real" use case (and hence code) would use malloc.

Many (majority of ?) embedded programmers would not use malloc. Moreover, some embedded programming standards (e.g. MISRA C) explicitly prohibit use of dynamic memory allocation. The more malloc's are in the Zephyr's codebase (even if in samples/), the stranger look Zephyr's aspirations to adhere to embedded programming best practices. Bottom line: unless the malloc is really due (e.g., there're samples for the POSIX subsystem, and malloc() is unalienable part of the POSIX), then not using malloc() is more "future-proof" (regarding where the Zephyr project may want to go).


My comment about that line: should be spaces around `+` in `length+1`.

lmajewski replied:
Ok. I will remove `malloc()` and replace it with the buffer automatically allocated.

jukkar commented on subsys/net/ip/connection.c:
```diff
@@ -601,8 +601,12 @@ enum net_verdict net_conn_input(struct net_pkt *pkt,
+		if ((IS_ENABLED(CONFIG_NET_UDP) ||
+		     IS_ENABLED(CONFIG_NET_TCP))
+#if defined(CONFIG_NET_DSA)
+		    && !net_dsa_recv(pkt_iface, pkt)
```
This call looks indeed in wrong place. It was a long time I last time reviewed this so have forgotten why do we actually need to have special handling for raw/packet sockets.

lmajewski replied:
**So the use case is as follows:**
We do want to have enabled in the same time support for TCP/IP processing (to e.g. have working `icmp` protocol)  and be able to handle some LLDP packets. The former shall be handled by Zephyr's network stack and the latter (LLDP) by threads spawn for each LAN{123} port (as it is presented in the dsa_lldp.c example).

For LLDP we use `socket(AF_PACKET, SOCK_RAW, ETH_P_ALL);` to create socket.
The code snippet above checks if received packet's MAC address matches the one which was written to KSZ8794's static table `01:80:c2:00:00:03`. The check is done in `dsa_ll_addr_switch_cb` (./samples/net/dsa/src/main.c) callback. If MAC
address marches - then the packet is NOT passed to Zephyr's network stack and is passed directly to LAN{123} interface (L687 in the same file) . 

This is "a hack", as the network stack shall assess if `AF_PACKET && SOCK_RAW` are used with socket creation and 
by itself pass packet to LAN{123} (this is at least how it works in Linux, from which the use case code was ported).
In the presented solution I do reuse the code, which sets the KSZ8794 static table and workaround lack of `AF_PACKET && SOCK_RAW` support with checking MAC address of incoming packets.

On the contrary, when I send the packet from lanX interfaces (`iface`) in the example code, the entry in static table forces using switch tag to forward the packet (with proper MAC) to the specified port. In that way I do achieve the most wanted feature - port separation between LAN{123} ports in the switch.

jukkar replied:
If you only want to catch LLDP packets, we already have functions for that. See `net_lldp_register_callback()` and friends in `include/net/lldp.h`

lmajewski replied:
LLDP is only the example - for which I was able to write a decent sample code.
In my use case - the work is performed on customized frames. In the past I've considered the code from ./samples/net/lldp/src/, but it did not fit.

lmajewski replied:
Could we discuss how this problem (i.e. lack of `AF_PACKET && SOCK_RAW`) could be solved? Any roadmap or possible obstacles could be helpful.

jukkar replied:
> Could we discuss how this problem (i.e. lack of `AF_PACKET && SOCK_RAW`) could be solved?

Yes, agreed. I have not forgotten this one :)

Naturally we need to support the combination of packet sockets and normal inet/inet6 sockets at the same time. After looking the code I think we could do this something like this:

* Modify/remove `packet_socket.c` so that packet is feed to connection.c without modifications. It seems that we might not need this file at all if the connection.c supports raw sockets properly.
* In `net_conn_input()`, when checking the recipient socket for the packet, check if there are any raw socket listeners and clone the packet (so that it can be delivered to multiple recipients). Similar thing is already done for multicast sockets (line 660), and for dgram type packet sockets (line 693).

The changes might actually be not that big. Would you be interest in implementing this?

lmajewski replied:
I will look on this closer after I restructure the code for DSA (i.e. put callbacks into proper structures).

jukkar reviewed (changes requested):
New review. One thing I noticed is that you had marked things as "Fixed" but had not changed the code.

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -647,7 +651,7 @@ static bool eth_get_ptp_data(struct net_if *iface, struct net_pkt *pkt)
 #endif /* CONFIG_PTP_CLOCK_MCUX */
 
-static int eth_tx(const struct device *dev, struct net_pkt *pkt)
+int eth_tx(const struct device *dev, struct net_pkt *pkt)
```
We should keep this as static, driver functions are suppose to be private.

lmajewski replied:
If I remember correctly - the last recommendation was to introduce something like:
`struct dsa_api` structure with `.send` member.

Then I would instantiate this structure in eth_mcux.c only when CONFIG_NET_DSA is defined and set
`send = eth_tx`

In that way eth_tx could stay `static` (private) in eth_mcux.c

jukkar replied:
> If I remember correctly - the last recommendation was to introduce something like:
> `struct dsa_api` structure with `.send` member.

That sounds good idea.

lmajewski replied:
**Another option would be to extend `struct ethernet_api`** from `./include/net/ethernet.h` to have:
`#if defined(CONFIG_NET_DSA)`
`	/** Send a network packet via DSA master port */`
`	int (*dsa_master_send)(const struct device *dev, struct net_pkt *pkt);`
`#endif`

The dsa_master_send callback could be set in `eth_mcux.c` and reused in DSA. From other comments you also proposed to add some MAC address related functions into the same structure.

**Yet another option would be to extend `struct ethernet_context`** in the same file to have
`/** Master ETH port's transmit callback - necessary for sending`
`* packets with DSA switch tag appended */`
`eth_tx_t master_eth_tx;`

Where: `typedef int (*eth_tx_t) (struct device *dev, struct net_pkt *pkt);`
and `eth_tx_t` corresponds to `eth_tx` from `eth_mcux.c`

**The question:**
I'm puzzled which place would be better. The `eth_ctx` is easier accessible from iface, but is harder to setup in `eth_mcux.c` (i.e. ETH driver files).

Any preferences?

lmajewski replied:
After some thoughts - IMHO it would be better to extend the struct ethernet_api as we will add some other callbacks (which e.g. set MAC address) there as well.

jukkar replied:
I was thinking more of a 3rd option:

Create `dsa_api` which extends `ethernet_api`. Then in `dsa_ksz8794.c`, instead of using `ethernet_api`, you would set
```
const struct dsa_api dsa_api_funcs = {
	.ethernet_api.iface_api.init	= dsa_iface_init,
	.ethernet_api.get_capabilities	= dsa_port_get_capabilities,
	.ethernet_api.send		= dsa_tx,
// add here all the dsa specific APIs you need
};
```
This way the driver would still be Ethernet compatible as it would have all the Ethernet callbacks in it, but we would avoid adding dsa specific stuff into ethernet_api struct.
Does this sound feasible?

lmajewski replied:
I've spent some time to provide implementation for this idea.
Unfortunately, it causes circular data dependencies between `./include/net/dsa.h` and `ethernet.h`.
The `struct ethernet_context` is using dsa callback exported in `./include/net/dsa.h` and `struct dsa_api` (from `dfsa.h`) uses `struct ethernet_api eapi;` defined in `ethernet.h`.

As fair as I can observe - the `ethernet.h` uses (and includes) data from other headers (like `net/lldp.h, net/ethernet_vlan.h`). In that way `struct ethernet_api` has callbacks for VLAN and PTP. 

What works with very little adjustments is to augment the `struct ethernet_api` with:
`#if defined(CONFIG_NET_DSA)`
`	/** DSA specific API callbacks */`
`	struct dsa_api dapi;`
`#endif`

And put all DSA related callbacks (also HW specific) into `struct dsa_api dapi`.

jukkar replied:
It there are circular dependencies, we would need to fix it then. But I know these can be tricky to solve.
Anyway, please propose then embedding dsa to ethernet_api, lets see how it looks like.

lmajewski replied:
Please find updated PR

lmajewski commented on drivers/ethernet/dsa_ksz8794.h:
```diff
@@ -0,0 +1,291 @@
+	KSZ8794_PORT1 = 1,
+	KSZ8794_PORT2,
+	KSZ8794_PORT3,
+	/* SWITCH <-> CPU port */
```
I made a mistake here. The port5 define was from KSZ8795 switch (with 4 LAN ports). I've fixed it now.

lmajewski commented on drivers/ethernet/Kconfig.dsa:
```diff
@@ -4,6 +4,15 @@
 
+menuconfig NET_DSA
+	bool "Distributed Switch Architecture support"
+	depends on NETWORKING
```
I see your point. Thanks for spotting this.
Indeed - IMHO, it would be better if DSA would depend on ETH_MCUX (as for now we only work with this driver).

If other architecture would like to add it - then one can add:
`       depends on ETH_MCUX || ETH_FOO`

tbursztyka commented on include/net/ethernet.h:
```diff
@@ -379,6 +396,15 @@ struct ethernet_context {
 	int8_t vlan_enabled;
 #endif
 
+#if defined(CONFIG_NET_DSA)
```
Put it above CONFIG_NET_VLAN and reverse the order of the attributes you added, for better alignment.

lmajewski replied:
Ok.

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,247 @@
+#define DSA_IS_PORT_MASTER(iface) \
+	(net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT)
+
+extern const struct ethernet_api dsa_api_funcs;
```
Could you be more specific? I do understand that you refer to `extern const struct ethernet_api dsa_api_funcs;` ?
This is needed as we do need to have LAN{123}  interfaces behaving as "normal" (i.e. non-DSA) ones.

The `struct ethernet_api's` address is then passed to `NET_SLAVE_DEVICE_INIT_INSTANCE`, which in turn is used with `DT_INST_FOREACH_CHILD(0, NET_SLAVE_DEVICE_INIT_INSTANCE);`

According to documentation - the  `NET_SLAVE_DEVICE_INIT_INSTANCE` shall accept only single argument, which would be the child number.

Maybe I've overlooked something, so any hints on expanding number of passed parameters is more than welcome (usage of `DT_INST_FOREACH_CHILD() )` seemed as a natural choice - but maybe there is a better way?.

lmajewski replied:
I think that I will move back the `#define NET_SLAVE_DEVICE_INIT_INSTANCE(slave)` again to `dsa_ksz8794.c`. In that way I can keep DSA related code in this file (and there will be NO need for extern declarations for functions)..

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,247 @@
+	NET_DEVICE_INIT_INSTANCE(dsa_slave_port_##slave,                   \
+	DT_LABEL(slave),                                                   \
+	0,                                                                 \
+	dsa_port_init,                                                     \
```
Please read the comment above for `extern const struct ethernet_api dsa_api_funcs;`

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,247 @@
+ */
+
+struct dsa_context {
+	uint8_t num_slave_ports;
```
Fixed.

lmajewski commented on subsys/net/ip/net_pkt.c:
```diff
@@ -974,6 +974,10 @@ static size_t pkt_buffer_length(struct net_pkt *pkt,
 			max_len += sizeof(struct net_eth_hdr);
+#if defined(CONFIG_NET_DSA)
+			/* Extend max pkt length for DSA to store tail tag.*/
+			max_len += _DSA_TAG_SIZE;
```
Ok. Fixed

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,247 @@
+
+/* DSA Macros */
+#define DSA_IS_PORT_MASTER(iface) \
+	(net_eth_get_hw_capabilities(iface) & ETHERNET_DSA_MASTER_PORT)
```
I've implicitly assumed that iface is an ethernet interface, yes.

lmajewski replied:
I've replaced it with a function.

lmajewski commented:
**Some notable changes:**
- The `ethernet_context` for master has the pointer to `dsa_context`
and `dsa_context` has dapi pointer to `dsa_api` structure

- Wrappers to get dsa_context introduced in dsa.{hc}

- The `ethernet_context` is not polluted by **any** dsa code - everything has been
moved to `dsa_context`.

- Rewrite `DSA_IS_PORT_MASTER()` from macro to static inline function - as it
also needs to take into consideration the check if iface is the ethernet one.

- Adjust naming in `dsa_ksz8794.c` file

**Now it looks like we _only_ need to focus on providing support for `AF_PACKET and SOCK_RAW` (i.e. remove https://github.com/zephyrproject-rtos/zephyr/pull/27620/commits/72d584abcebb99748b8c6aa5a37ac71b3915e45c hack)**

jukkar commented on subsys/net/ip/net_core.c:
```diff
@@ -393,6 +396,9 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	net_pkt_cursor_init(pkt);
 
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
```
Wouldn't you be able to call this in the driver so we would not need to call this in the generic function?

lmajewski replied:
After some excessive testing it turned out that the `dsa_recv_set_iface()` shall
be called in `net_recv_data()` (subsys/net/ip/net_core.c) as after some basic
checks (if e.g. `iface != NULL`) we are sure that the iface is ready to handle
the packet. To achieve the same results - one would need to copy those checks to `eth_rx()`
(defined in eth_mcux.c) which IMHO is the overkill.

Moreover, in the same function the code to assign original `iface` (when
`CONFIG_NET_ROUTING` is enabled) is executed and hence the deduction that
DSA shall follow the same pattern.

lmajewski replied:
The culprit was the `overwrite` flag being not set in `struct pkt`. The fix was to call `net_pkt_set_overwrite(pkt, true);` before doing any tag related manipulations.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,202 @@
+	return false;
+}
+#else
+bool dsa_is_port_master(struct net_if *iface)
+{
+	return false;
+}
```
This will need to be in header file. If Ethernet is not enabled, then this file should not be compiled at all so the check would not do anything here.

lmajewski replied:
The only place where `dsa_is_port_master()` is used outside DSA related code (i.e. protected with `CONFIG_NET_DSA`) is
the packet_socket.c file. However, this file is going to be removed (as suggested in the other review reply) and the `connection.c` is going to be rewritten to avoid usage of this function.

(The declaration of `dsa_is_port_master()` is already placed in `include/net/dsa.h` file)

jukkar commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,202 @@
+	return 0;
+}
+
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt)
```
So basically this "hijacks" the packet. I think we should leave `net_recv_data()` as is and you could add all the DSA specific stuff to `ethernet.c`.

lmajewski replied:
I'm not sure what do you mean by packet "hijacks" . The `dsa_recv_set_iface()` is responsible for:
1. Use `dsa_ksz8794_get_iface()` to read the tag and return proper `struct net_if` pointer to LAN{123} interfaces. That is why we need to call it early to have the tag data (from switch IC) not removed and packet interface set before it is used by network stack. In that way the packet shall be passed to proper LAN{123} interface.
2. (optionally) - The `net_dsa_recv()` function calls the user application registered callback (in the ./samples/net/dsa case it is `dsa_ll_addr_switch_cb()` which just checks the `lladst.addr`).

Please also note that in the end of `net_recv_data()` (line:427) we call `net_pkt_set_iface(pkt, iface);` which assigns iface to packet. After that the packet is passed to proper RX queues.

I could add the `dsa_recv_set_iface()` to `ethernet_recv()` but IMHO considering the above arguments the current placement of this function is optimal.

jukkar replied:
I did quick experiment and moved the stuff from net_core.c to mcux driver. Why would this not work?
```
diff --git a/drivers/ethernet/eth_mcux.c b/drivers/ethernet/eth_mcux.c
index 8d13bfa1b8..2b85278a67 100644
--- a/drivers/ethernet/eth_mcux.c
+++ b/drivers/ethernet/eth_mcux.c
@@ -710,6 +710,7 @@ static void eth_rx(struct eth_context *context)
 {
 	uint16_t vlan_tag = NET_VLAN_TAG_UNSPEC;
 	uint32_t frame_length = 0U;
+	struct net_if *iface;
 	struct net_pkt *pkt;
 	status_t status;
 	unsigned int imask;
@@ -811,7 +812,13 @@ static void eth_rx(struct eth_context *context)
 
 	irq_unlock(imask);
 
-	if (net_recv_data(get_iface(context, vlan_tag), pkt) < 0) {
+	iface = get_iface(context, vlan_tag);
+
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
+#endif
+
+	if (net_recv_data(iface, pkt) < 0) {
 		net_pkt_unref(pkt);
 		goto error;
 	}
diff --git a/subsys/net/ip/net_core.c b/subsys/net/ip/net_core.c
index c0a63c7172..f60bec1fad 100644
--- a/subsys/net/ip/net_core.c
+++ b/subsys/net/ip/net_core.c
@@ -396,9 +396,6 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	net_pkt_set_overwrite(pkt, true);
 	net_pkt_cursor_init(pkt);
 
-#if IS_ENABLED(CONFIG_NET_DSA)
-	iface = dsa_recv_set_iface(iface, &pkt);
-#endif
 	NET_DBG("prio %d iface %p pkt %p len %zu", net_pkt_priority(pkt),
 		iface, pkt, net_pkt_get_len(pkt));
```

lmajewski replied:
This will not work as you can pass `iface == NULL` to `dsa_recv_set_iface()` as the argument. This function is not checking it as such checks are already done in `net_recv_data()`. 
As a result - some allocation errors pops up and the stack is not usable anymore.

Moreover, the `net_pkt_set_overwrite(pkt, true);` is executed `net_recv_data()` as well, so packet manipulation is possible (to remove tag).

jukkar reviewed (commented):
There was lot of code to review so I might have missed some, anyway looks much better now.

jukkar commented on boards/arm/ip_k66f/ip_k66f.dts:
```diff
@@ -95,6 +95,7 @@
 
 &enet {
 	status = "okay";
+	local-mac-address = [00 00 12 13 00 10];
```
Why do we have this here, wouldn't all devices have this MAC by default now?

lmajewski replied:
This was for the LLDP example. In this case `local-mac-address` can be observed when I do test the LLDP example for this PR.
As a test - I use `lldpd` from Linux, but simultaneously I do ping the eth device. In that way I can check if network stack is working properly when with LLDP I test `SOCK_RAW and AF_PACKET`.

jukkar replied:
> This was for the LLDP example.

If this was for one sample, then we should not have it set here in the board dts file. You can add dt overlays in the sample directory to have these values there.

lmajewski commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,1038 @@
+		 * Provide pointer to DSA context to master's eth interface
+		 * struct ethernet_context
+		 */
+		ctx_master = net_if_l2_data(context->iface_master);
```
Yes, in the above code, which checks `context->iface_master` the `return` is missing. I will add it.

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,227 @@
+ */
+
+struct dsa_context {
+	struct net_if *iface_slave[NET_DSA_PORT_MAX_COUNT];
+	struct net_if *iface_master;
+	struct k_delayed_work dsa_work;
```
Thanks for the tip with `INTERNAL_HIDDEN`. I've added missing descriptions.

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,227 @@
+ * api callbacks
+ */
+struct dsa_switch_hw_api {
+	int w;
```
This shall be removed as it is a leftover from some earlier re-work of the code (the idea was to put all "switch" related code to this structure, but I think that too many levels of structures is not necessary).

lmajewski commented on include/net/dsa.h:
```diff
@@ -0,0 +1,227 @@
+ */
+
+struct dsa_slave_config {
+	uint8_t mac_addr[6];
```
Added. This field is to specify separate MAC addresses per LAN{123} interfaces.

lmajewski commented on include/net/ethernet.h:
```diff
@@ -371,6 +391,22 @@ struct ethernet_context {
+	/** DSA RX callback function - for custom processing - like e.g.
+	 * redirecting packets when MAC address is caught
+	 */
+	net_dsa_recv_cb_t cb;
```
Ok. I've renamed both to be more descriptive.

Commit:
dts: overlay: Provide local-mac-address nodes as dts overlay

The 'local-mac-address' nodes are helpful for assessing if the LLDP like
example of DSA switch works correctly.

As those are used only for testing (or when user explicitly needs them),
those are added as DTS overlay to DSA example.

The 'local-mac-address' properties will be visible when one assess the
LLDP response from ip_k66f board (each LAN port has different IP
assigned).

Example output from 'tcpdump -v -i eth0':
-----------------------------------------
       Chassis ID TLV (1), length 7
          Subtype MAC address (4): 00:00:12:13:00:37 (oui Unknown)
        Port ID TLV (2), length 7
          Subtype MAC address (3): 00:00:12:13:00:37 (oui Unknown)
        Time to Live TLV (3), length 2: TTL 120s
        System Name TLV (5), length 13: ip_k66f LAN:3
        End TLV (0), length 0


Signed-off-by: Lukasz Majewski <lukma@denx.de>

lmajewski commented:
This PR depends on https://github.com/zephyrproject-rtos/zephyr/pull/30934 which allows usage of raw sockets also when support for TCP and UDP is enabled.

jukkar commented on drivers/ethernet/dsa_ksz8794.c:
```diff
@@ -0,0 +1,1040 @@
+}
+#endif
+
+void dsa_iface_init(struct net_if *iface)
```
This should probably be static

lmajewski replied:
Yes. Thanks for pointing this out.

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -807,7 +812,11 @@ static void eth_rx(struct eth_context *context)
 	irq_unlock(imask);
 
-	if (net_recv_data(get_iface(context, vlan_tag), pkt) < 0) {
+	iface = get_iface(context, vlan_tag);
+#if IS_ENABLED(CONFIG_NET_DSA)
+	iface = dsa_recv_set_iface(iface, &pkt);
+#endif
```
Have you tried to use DSA with VLAN?
I am not sure about the ordering of these two lines, just wondering should the `get_iface()` call be after the dsa one. It really depends how the VLAN master interface is setup (is it the DSA interface or one of the DSA slave interfaces).

lmajewski replied:
No, I did not tested it as my use case don't require and use VLANs. I just rely on the switch IC (KSZ8794) appended tag.

From the code in `drivers/ethernet/eth_mcux.c` it  seems like the VLAN related fields (TPID, VID) are only read (and the packet itself is not modified), so the tag will stay in place (appended to the end of the packet).

It looks to me that (by default) DSA shall be used when VLAN is disabled, otherwise (as you mentioned) there may be an issue with selected `iface`. This is not a problem, as user application may use registered `ctx->dsa_recv_cb` callback to just return 0 in `dsa_recv_set_iface()`, so the VLAN iface is used.

jukkar commented on include/net/dsa.h:
```diff
@@ -0,0 +1,222 @@
+	/** Status of each port */
+	bool link_up[NET_DSA_PORT_MAX_COUNT];
+
+	/** DSA specific API callbacks - filled in the switch IC driver */
+	struct dsa_api *dapi;
```
Minor nit, but could you place this pointer after the `iface_master` one, there is less chance of leaving holes in the struct that way.

jukkar reviewed (approved):
Looks good, couple of minor nits.

jukkar commented:
We could try to get this to 2.5 if possible. This one depends on #30934 which should also get in to 2.5

mbolivar-nordic commented on dts/bindings/dsa/microchip,ksz8794.yaml:
```diff
@@ -0,0 +1,63 @@
+        required: false
+        description: |
+          Define the Low-Speed Interface Drive Strength for MII and RMMI
+          Supported values 2,4,8,12,16,20,24,28mA
```
This should use an `enum`, then.

lmajewski replied:
Ok. Fixed Thanks for pointing this out.

mbolivar-nordic replied:
> Ok. Fixed Thanks for pointing this out.

Did you forget to push?

lmajewski replied:
I've just pushed updated patches

mbolivar-nordic replied:
This `enum` looks OK. Thanks.

mbolivar-nordic commented on dts/bindings/dsa/microchip,ksz8794.yaml:
```diff
@@ -0,0 +1,76 @@
+        type: int
+        required: false
+        description: |
+          Define the applied workaround for the switch used for
+          short connections. Use bitmask to select the workaround or more
+          0x01: Short Cable Problems with the KSZ8795 Family
+          0x02: 1) CAT-5E/6 Short Cable with a Link Issue for the KSZ8795 Family
+          0x04: 2) CAT-5E/6 Short Cable with a Link Issue for the KSZ8795 Family
+        enum:
+          - 0x01
+          - 0x02
+          - 0x04
```
Hmm, I wasn't asking for this to be an `enum`. The documentation says it's a bitmask, which makes it look like 0x7 is a valid value. Using an `enum` would cause an error if the user tries that though. Are you sure?

lmajewski replied:
Ok, I've re-check it. In this particular case - the "workaround" property shouldn't be an enum, as it is valid to have 0x7 as the value. This enum shall be removed.
However, IMHO it looks correct to have enum in the 'mii-lowspeed-drivestrength' property, as set current values are distinct.

tbursztyka commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -0,0 +1,209 @@
+	return 0;
+}
+
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt)
```
I would just merge the content of that function into dsa_net_recv().

Now that you got rid of using dsa_net_revc from net_recv_data, I think you can just call dsa_net_recv in relevant drivers directy (with the iface and the **pkt).

It's seems more logical to do it that way.

I think it's my last comment. rest looks good to me

lmajewski replied:
Please correct me if I've misunderstood you:
You propose to:
1. Rename `dsa_recv_set_iface` (line 126 above) to `dsa_net_recv`
2. The content of `static int dsa_net_recv(struct net_if *iface, struct net_pkt *pkt)` (line 108) merge into the renamed function from point 1 above.

Am I correct?

lmajewski replied:
I've updated the code - please check

mbolivar-nordic commented on dts/bindings/dsa/microchip,ksz8794.yaml:
```diff
@@ -0,0 +1,72 @@
+          - 28
+
+child-binding:
+    description: Praperties of slave port
```
```suggestion
    description: Properties of slave port
```

mbolivar-nordic reviewed (approved):
DT bits look good. I did find one typo I would like to see fixed in the child binding description.

jukkar commented:
@nashif could you merge this? There is false positive from checkpatch that is blocking the merge atm.

```
:1144: WARNING:TRAILING_SEMICOLON: macros should not use a trailing semicolon
#1144: FILE: drivers/ethernet/dsa_ksz8794.c:1023:
+#define NET_SLAVE_DEVICE_INIT_INSTANCE(slave)                              \
+	const struct dsa_slave_config dsa_0_slave_##slave##_config = {     \
+		.mac_addr = DT_PROP_OR(slave, local_mac_address, {0})      \
+	};                                                                 \
+	NET_DEVICE_INIT_INSTANCE(dsa_slave_port_##slave,                   \
+	DT_LABEL(slave),                                                   \
+	0,                                                                 \
+	dsa_port_init,                                                     \
+	device_pm_control_nop,                                             \
+	&dsa_context,                                                      \
+	&dsa_0_slave_##slave##_config,                                     \
+	CONFIG_ETH_INIT_PRIORITY,                                          \
+	&dsa_eth_api_funcs,                                                \
+	ETHERNET_L2,                                                       \
+	NET_L2_GET_CTX_TYPE(ETHERNET_L2),                                  \
+	NET_ETH_MTU);

- total: 0 errors, 1 warnings, 2655 lines checked
```

jukkar commented:
@lmajewski what changes did you do to fix the checkpatch issue?
